### PR TITLE
Change Drivetrain Deadband

### PR DESCRIPTION
**Pull request by lesereli000**
Adds a couple of changes to the deadband on drivetrain:
<ul>
<li>add a lowered power cap</li>
<li>add a "boost" button to raise power cap</li>
<li>change algorithm for determining power to be non-linear</li>
<li>add a translation to the curve so that when the deadband is no longer active there is no jump in power</li>
</ul>

### CODE REVIEW

File: src/main/java/frc/robot/Drivetrain.java [hunk 1]
```diff
@@ -28,8 +28,8 @@
      * 
      * @param speed the speed at which to drive (ranges from -1.0 to +1.0)
      */
-    public void driveLeft(double speed){
-        double sp = deadband(speed);
+    public void driveLeft(double speed, boolean boost){
```
Comment (tricktrap) [inline src/main/java/frc/robot/Drivetrain.java]: I like the idea of boost, but I think we should move the implementation "up" to avoid making autonomous code have to deal with it. I recommend moving it outside of the class entirely, and only using it in the teleopPeriodic functions.

File: src/main/java/frc/robot/Drivetrain.java [hunk 2]
```diff
@@ -75,9 +75,11 @@ public void tankDrive(double leftDrive, double rightDrive){
      * motors.
      *
      * @param in the input to check
+     * @param boost whether or not the boost button is being pressed 
      * @return 0.0 if {@code in} is less than abs(DEADBAND_LIMIT) else {@code in}
      */
-    public double deadband(double in) {
-        return Math.abs(in) > DEADBAND_LIMIT ? in : 0.0;
+    public double deadband(double in, boolean boost) {
```
Comment (tricktrap) [inline src/main/java/frc/robot/Drivetrain.java]: This function is actually doing a little too much at this point. It's doing both "deadband" and "boost" functions which should probably be separated into their own functions. Once you do so, `deadband` doesn't need to be in the drivetrain class at all, in fact it might be useful to keep separate so that other classes can use it as well (manual control of the shooter angle, collection system, etc.) 

The boost code itself is pretty hard to follow, I think we should move all instances of `0.6` to a constant (also to make it easy to change later). Doubling up the terenary operators compacts the code but makes it harder to read and debug, it's worth having extra whitespace to make this easier to read.

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -160,7 +160,7 @@ public void teleopPeriodic() {
         limelight.setLightEnabled(false);
       }
 
-      drive.arcadeDrive(turnInput, speedInput);
+      drive.arcadeDrive(turnInput, speedInput, driver.getAButtonPressed());
```
Comment (tricktrap) [inline src/main/java/frc/robot/Robot.java]: I think this is the best place to call the boost code (hopefully as a function). Rather than pass whether to boost down two or more layers of function calls, we can just adjust turnInput and speedInput here instead.

**Review by tricktrap — CHANGES_REQUESTED** [review]
Ok, so the intent and final result of this code is good (it will almost certainly work), but I think it should be organized a little differently to be more maintainable.

File: src/main/java/frc/robot/Drivetrain.java (hunk 1 above)
Comment (lesereli000) [reply src/main/java/frc/robot/Drivetrain.java]: boost is now called from tank drive and arcade drive methods. 

public void tankDrive(double leftDrive, double rightDrive, boolean boost){
        this.driveLeft(boost(leftDrive, boost));
        this.driveRight(boost(rightDrive, boost));
    }

File: src/main/java/frc/robot/Drivetrain.java (hunk 2 above)
Comment (lesereli000) [reply src/main/java/frc/robot/Drivetrain.java]: methods are separated and deadband is in robot class

File: src/main/java/frc/robot/Robot.java (hunk 3 above)
Comment (lesereli000) [reply src/main/java/frc/robot/Robot.java]: implemented

**Review by tricktrap — APPROVED** [review]
Even though I totally hijacked this PR I still approve

**Review by sreeves750 — APPROVED** [review]
Looks good. All that's left is to throw it on the drivetrain and make sure it works.

**Review by TheBitEffect — APPROVED** [review]
It compiles! \\(^◇^)/

**Review by dracco1993 — APPROVED** [review]
